### ISSUE TITLE
Installation issue:  py-jupyter-server build error when using different gcc backend for Intel

### ISSUE THREAD

**Issue by climbfuji** [LLNL/spack#41899]
### Steps to reproduce the issue

Mentioning @adamjstewart as per instructions.

The problem occurs when building `py-jupyter-server` on a system with Intel when the `gcc` backend (`libstdc++`) is configured to be different from the default OS `gcc`. The reason is that the build of `py-jupyter-server` invokes `node12` (although it's not configured as a dependency for the package), which if installed as an OS package, then looks for certain `GLIBCXX` strings in the OS `/usr/lib64/libicuuc.so.*` library.

I tried to duct-tape this by simply makein `icu4c` a dependency for `py-jupyter-server` and hoping that when it gets loaded `node12` finds it, but that didn't work.

This may also be an issue when using a different `gcc` than the default OS `gcc`, if the `libstdc++` libraries are different enough.

### Error message

```
  Added file:///p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/cache/build_stage/heinzell/spack-stage-py-jupyter-server-1.21.0-d6gzxy2bdlcdbnjzucznopwk2jjqge7l/spack-src to build tracker '/p/work1/tmp/heinzell/pip-build-tracker-zjtkf504'
  Created temporary directory: /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4
  Preparing metadata (pyproject.toml): started
  Running command Preparing metadata (pyproject.toml)
  running dist_info
  creating /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info
  writing /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/PKG-INFO
  writing dependency_links to /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/dependency_links.txt
  writing entry points to /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/entry_points.txt
  writing requirements to /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/requires.txt
  writing top-level names to /p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/top_level.txt
  writing manifest file '/p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/SOURCES.txt'
  reading manifest file '/p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no previously-included files found matching 'docs/#*'
  no previously-included directories found matching 'docs/build'
  no previously-included directories found matching 'docs/gh-pages'
  no previously-included directories found matching 'docs/dist'
  warning: no previously-included files matching '*~' found anywhere in distribution
  warning: no previously-included files matching '*.pyo' found anywhere in distribution
  no previously-included directories found matching '.git'
  no previously-included directories found matching '**/.ipynb_checkpoints'
  no previously-included directories found matching '**/.pytest_cache'
  no previously-included directories found matching '**/.coverage'
  no previously-included directories found matching '**/.pytest_cache'
  adding license file 'COPYING.md'
  writing manifest file '/p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server.egg-info/SOURCES.txt'
  creating '/p/work1/tmp/heinzell/pip-modern-metadata-kmq_o3i4/jupyter_server-1.21.0.dist-info'
  Preparing metadata (pyproject.toml): finished with status 'done'
  Source in /p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/cache/build_stage/heinzell/spack-stage-py-jupyter-server-1.21.0-d6gzxy2bdlcdbnjzucznopwk2jjqge7l/spack-src has version 1.21.0, which satisfies requirement jupyter-server==1.21.0 from file:///p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/cache/build_stage/heinzell/spack-stage-py-jupyter-server-1.21.0-d6gzxy2bdlcdbnjzucznopwk2jjqge7l/spack-src
  Removed jupyter-server==1.21.0 from file:///p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/cache/build_stage/heinzell/spack-stage-py-jupyter-server-1.21.0-d6gzxy2bdlcdbnjzucznopwk2jjqge7l/spack-src from build tracker '/p/work1/tmp/heinzell/pip-build-tracker-zjtkf504'
Created temporary directory: /p/work1/tmp/heinzell/pip-unpack-_ipfugv_
Building wheels for collected packages: jupyter-server
  Created temporary directory: /p/work1/tmp/heinzell/pip-wheel-j70n0wce
  Destination directory: /p/work1/tmp/heinzell/pip-wheel-j70n0wce
  Building wheel for jupyter-server (pyproject.toml): started
  Running command Building wheel for jupyter-server (pyproject.toml)
  /usr/bin/node12: /opt/cray/pe/gcc/10.3.0/snos/lib64/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /usr/lib64/libicuuc.so.suse65.1)
  Traceback (most recent call last):
    File "/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/py-pip-23.1.2-5yxaluf/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
      main()
    File "/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/py-pip-23.1.2-5yxaluf/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/py-pip-23.1.2-5yxaluf/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 251, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
    File "/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/py-jupyter-packaging-0.12.0-unzl5oh/lib/python3.10/site-packages/jupyter_packaging/build_api.py", line 24, in build_wheel
      builder()
    File "/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/py-jupyter-packaging-0.12.0-unzl5oh/lib/python3.10/site-packages/jupyter_packaging/setupbase.py", line 230, in builder
      run(npm_cmd + ["install"], cwd=node_package)
    File "/p/app/projects/NEPTUNE/spack-stack/spack-stack-1.6.0/envs/unified-env-intel-2021.4.0/install/intel/2021.4.0/py-jupyter-packaging-0.12.0-unzl5oh/lib/python3.10/site-packages/jupyter_packaging/setupbase.py", line 296, in run
      return subprocess.check_call(cmd, **kwargs)
...
```

### Information on your system

```
> spack debug report
* **Spack:** 0.21.0 (4c6dcd2a01927d06ccc1005d8403811d0eccdc30)
* **Python:** 3.9.7
* **Platform:** linux-sles15-zen2
* **Concretizer:** clingo
```

### Additional information

[spack-build-01-install-out.txt](https://github.com/spack/spack/files/13796520/spack-build-01-install-out.txt)
[spack-build-env-mods.txt](https://github.com/spack/spack/files/13796521/spack-build-env-mods.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/13796522/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/13796523/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by climbfuji** [comment]
@adamjstewart One workaround I found was to install node.js manually outside of spack with the correct gcc compiler loaded. That was surprisingly simple and allowed the build to proceed (just make sure the spack compiler config loads the node.js module).

**Comment by adamjstewart** [comment]
Should we add node.js as a dependency? Does that fix the issue?

**Comment by climbfuji** [comment]
> Should we add node.js as a dependency? Does that fix the issue?

I assume this will work. My quick-and-dirty approach was to simply add the `node.js` paths into the compiler config on the system. I was a bit hesitant making it a dependency because `node.js` doesn't build out of the box in spack (where I tried it), but of course an external package should do.

**Comment by adamjstewart** [comment]
Hmm, I'm able to build py-jupyter-server 2.6.0 and 1.21.0 without node/npm, so we definitely don't want to make it a required dep. Maybe it's only used if it's found? Do you have the same issue with py-jupyter-server 2.6.0?

**Comment by climbfuji** [comment]
No, 2.6.0 builds fine. The package I am after `py-cylc-uiserver` has a constraint on `py-jupyter-server<2` in the package config. I removed that for testing if `py-cylc-uiserver` worked fine with `py-jupyter-server@2.6.0`, but unfortunately it didn't.

**Comment by adamjstewart** [comment]
Looks like that constraint was dropped in py-cylc-uiserver 1.4 if you want to update that recipe.

**Comment by adamjstewart** [comment]
The best alternative is to patch out whatever calls node in jupyter-server 1

**Comment by climbfuji** [comment]
Thanks for your help. Let's keep this open for now. I've got a release to roll out where I am using my duct-tape solution to get it done. After that I'll fix it properly and submit a PR to spack develop.

**Comment by climbfuji** [comment]
@adamjstewart I finally took the time to sort this out. The story behind this is: `py-jupyter-server@:1` uses `py-jupyter-packaging` as build tool (v2 uses something else).

There is an optional feature to build the `typescript` extension when `npm` is found, but unfortunately this is all hardcoded to be auto-detected in the `py-jupyter-server` build and the underlying `py-jupyter-packaging` tool. The CLI of `py-jupyter-packaging` provides an option to skip npm, but this is not available when you import the Python modules as `py-jupyter-server` does.

I found a way around this, which is to add a variant `typescript` that is `False` by default. For `~typescript`, I am patching (hacking?) `pyproject.toml` to pass a nonsense build argument `npm=/dev/null` to the `npm_builder`, which means it can't find it and gives up. For `+typescript`, I am adding the `npm` dependency explicitly to the package.

I tested both installs `+/~typescript` on the two systems that had errors with `py-jupyter-server` (the error that lead to the creation of this issue, and also a weird macOS where `npm` was found, but `node` was not), and they all worked.

This is all in PR https://github.com/spack/spack/pull/43279
